---
title: "Transcriptome Demo"
author: "Tricia"
date: "`r Sys.Date()`"
output: pdf_document
---

# Load required packages (you might have to figure out how to install some of these first...)
```{r message=FALSE}
library(ballgown)
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
```

# Setting up a dataset that associates the plankton files with planktonic and the biofilm files with biofilm labels
```{r}
pheno_data<-data.frame(ids = c("plank01", "plank02", "biofilm01", "biofilm02"),
                       stage = c("planktonic", "planktonic", "biofilm", "biofilm"))
```

# create Ballgown object and check transcript number
```{r message=FALSE}
samples.c <- paste('ballgown', pheno_data$ids, sep = '/')
bg <- ballgown(samples = samples.c, meas='all', pData = pheno_data)
bg
```

# This code is filtering the ballgown object and only keeping data that is dissimilar, eliminating repeats 
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt
```

# create a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)
```

# choose a transcript to examine more closely (this is a demo, you need to choose another)
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-PA0004", ]
```
## what information are you given about this transcript? Its name is gyrB, its p value is 0.274308233 and its q value is 0.9471885

# This code filters the transcripts out that have a p value of greater than 0.05 
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)
```

# organize the table by pval and the absolute value of fc. These are put in ascending order
```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)
```

# load gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

# pull out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```

# it is renaming the table so that the FPKM. prefix is removed
```{r}
colnames(gene_expression) <- c("plank01", "plank02", "biofilm01", "biofilm02")
head(gene_expression)
dim(gene_expression)
```

# load the transcript to gene table and determine the number of transcripts and unique genes
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))
```

# plot the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

## There is a high number of genes with a single transript. As the number of transcripts per gene increases, the frequency of those genes decreases. I believe this means each gene codes for a single transcript, on average.

# create a plot of how similar the two replicates are for one another. We have two data sets...how can you modify this code in another chunk to create a plot of the other set?
```{r}
x = gene_expression[,"plank01"]
y = gene_expression[,"plank02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="blue", cex=0.25,
xlab="FPKM (Planktonic, Replicate 1)", ylab="FPKM (Planktonic, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "black")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")

x = gene_expression[,"biofilm01"]
y = gene_expression[,"biofilm02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="red", cex=0.25,
xlab="FPKM (Biofilm, Replicate 1)", ylab="FPKM (Biofilm, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "black")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

## What does it mean if the two data sets are similar? It would mean that they have similar expression values of the same gene. Possibly meaning the gene of interest has an essential function in both the planktonic and biofilm bacteria

# create plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen",
xlab="log2(Fold change) Planktonic vs Biofilm",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```

## interpret the above figure: There appears to be a lot of differential expression, because there is a high frequency above and below 2. Values below -2 represent a higher expression in the biofilm, while values above 2 represent higher expression in planktonic

# Plot total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"plank"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"biofilm"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"plank"]+min_nonzero)
y=log2(gene_expression[,"biofilm"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Planktonic FPKM (log2)", ylab="Biofilm FPKM (log2)",
main="Planktonic vs Biofilm FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="red", pch=16, cex=0.5)
legend("topleft", "Significant", col="green", pch=16)
```

# make a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```

# choose a gene to determine individual expression gene 4
```{r}
ballgown::transcriptNames(bg_filt)[4]
ballgown::geneNames(bg_filt)[4]
```
# transform to log2
```{r}
transformed_fpkm <- log2(fpkm[4, ] + 1)
```

# make sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

# plot expression of individual gene
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[4], ' : ', ballgown::transcriptNames(bg_filt)[4]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("lightblue", "salmon"),
        border="darkblue")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```

## interpret the above figure: the gene PA0004 is expressed at a higher level in the planktonic species than in the biofilm species. However, the biofilm species has a higher variation than the plantonic species.